processor/logdeduplicationprocessor/counter.go

```diff
@@ -0,0 +1,165 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package logdeduplicationprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/logdeduplicationprocessor"
+
+import (
+	"hash/fnv"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil"
+)
+
+// Attributes names for first and last observed timestamps
+const (
+	firstObservedTSAttr = "first_observed_timestamp"
+	lastObservedTSAttr  = "last_observed_timestamp"
+)
+
+// timeNow can be reassigned for testing
+var timeNow = time.Now
+
+// logAggregator tracks the number of times a specific logRecord has been seen.
+type logAggregator struct {
+	resources         map[[16]byte]*resourceAggregator
+	logCountAttribute string
+	timezone          *time.Location
+}
+
+// newLogAggregator creates a new LogCounter.
+func newLogAggregator(logCountAttribute string, timezone *time.Location) *logAggregator {
+
+	return &logAggregator{
+		resources:         make(map[[16]byte]*resourceAggregator),
+		logCountAttribute: logCountAttribute,
+		timezone:          timezone,
+	}
+}
+
+// Export exports the counter as a Logs
+func (l *logAggregator) Export() plog.Logs {
+	logs := plog.NewLogs()
+
+	for _, resource := range l.resources {
+		resourceLogs := logs.ResourceLogs().AppendEmpty()
+		resourceAttrs := resourceLogs.Resource().Attributes()
+		resourceAttrs.EnsureCapacity(resourceAttrs.Len())
+		resource.attributes.CopyTo(resourceAttrs)
+
+		scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+		for _, lc := range resource.logCounters {
+			lr := scopeLogs.LogRecords().AppendEmpty()
+
+			baseRecord := lc.logRecord
+
+			// Copy contents of base record
+			baseRecord.SetObservedTimestamp(pcommon.NewTimestampFromTime(lc.firstObservedTimestamp))
+			baseRecord.Body().CopyTo(lr.Body())
+
+			lr.Attributes().EnsureCapacity(baseRecord.Attributes().Len())
+			baseRecord.Attributes().CopyTo(lr.Attributes())
+
+			lr.SetSeverityNumber(baseRecord.SeverityNumber())
+			lr.SetSeverityText(baseRecord.SeverityText())
+
+			// Add attributes for log count and timestamps
+			lr.SetObservedTimestamp(pcommon.NewTimestampFromTime(lc.firstObservedTimestamp))
+			lr.SetTimestamp(pcommon.NewTimestampFromTime(timeNow()))
+			lr.Attributes().PutInt(l.logCountAttribute, lc.count)
+
+			firstTimestampStr := lc.firstObservedTimestamp.In(l.timezone).Format(time.RFC3339)
+			lastTimestampStr := lc.lastObservedTimestamp.In(l.timezone).Format(time.RFC3339)
+			lr.Attributes().PutStr(firstObservedTSAttr, firstTimestampStr)
+			lr.Attributes().PutStr(lastObservedTSAttr, lastTimestampStr)
+		}
+	}
+
+	return logs
+}
+
+// Add adds the logRecord to the resource aggregator that is identified by the resource attributes
+func (l *logAggregator) Add(resourceAttrs pcommon.Map, logRecord plog.LogRecord) {
+	key := pdatautil.MapHash(resourceAttrs)
```

Comment:
Since `Add` is typically called repeatedly for each resource, we should hash the resource attributes in the caller and pass them in. Especially important because all of this is within a mux.